### PR TITLE
Refactor Expr classes API (part 10)

### PR DESCRIPTION
**Pull request by ienkovich**
This part introduces new visitors for Expr. The reason for that is to simplify some visitors and make them more efficient. The current visitor has the inconvenience of having all virtual methods being const, has no opportunity of `void` return type, and creates a lot of unnecessary objects for `aggregateResult`. I introduce the new visitor `ExprVisitor` and its specialization `ExprCollector` which implements a widely used template for collecting some data from an expression tree. The old visitor still lives because it fits a few of the current use cases.

### CODE REVIEW

File: IR/ExprVisitor.h [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+  virtual T visitAggExpr(const hdk::ir::AggExpr* agg) {
+    if (agg->arg()) {
+      return visit(agg->arg());
+    }
+    return defaultResult(agg);
+  }
+
+  virtual T defaultResult(const hdk::ir::Expr*) const {
+    if constexpr (!std::is_same<T, void>::value) {
+      return T{};
```
Comment (kurapov-peter) [inline IR/ExprVisitor.h:304]: What is this return value for?

**Review by kurapov-peter — APPROVED** [review]
Looks good, just a couple questions.

File: IR/ExprCollector.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  ResultType& result() { return result_; }
+  const ResultType& result() const { return result_; }
+
+ protected:
+  ResultType result_;
+};
+
+}  // namespace hdk::ir
```
[the remark this answers is not in the document]
Comment (ienkovich) [reply IR/ExprCollector.h]: Fixed

File: QueryEngine/ExprByPredicateCollector.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+class ExprByPredicateCollector
+    : public hdk::ir::ExprCollector<std::list<const hdk::ir::Expr*>,
+                                    ExprByPredicateCollector> {
+ public:
+  ExprByPredicateCollector(bool (*pred)(const hdk::ir::Expr*)) : pred_(pred) {}
+
+  void visit(const hdk::ir::Expr* expr) override {
+    if (pred_(expr)) {
+      if (std::find(result_.begin(), result_.end(), expr) == result_.end()) {
```
[the remark this answers is not in the document]
Comment (ienkovich) [reply QueryEngine/ExprByPredicateCollector.h:17]: There are usages when the first element of the returned list is used, so the order might matter. I'm not sure that code is correct though, but decided to keep the original behavior.

File: IR/ExprVisitor.h (hunk 1 above)
Comment (ienkovich) [reply IR/ExprVisitor.h:304]: Some visitors use returned values, e.g. ExprRewritter.

**Pull request merged**